### PR TITLE
NEP: Making details of ``np.core`` private

### PR DESCRIPTION
**Pull request by eric-wieser**
This is a rough draft of what I had in mind in the comments of #11909.

I don't think I'll have the energy to champion this, and would greatly appreciate if someone else could take over from me

### CODE REVIEW

File: doc/neps/nep-0023-restructuring-np-core.rst [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    _exposed = set(globals()) - _globals
+    assert _exposed == set(__all__) | set(_old_dir)
+
+
+Something to be aware of here will be the fact that ``dir`` of ``np.core.numerictypes``, and of any module that does ``from np.core.numerictypes import *`` is platform-dependent, as it contains all the type aliases.
+
+Backward compatibility
+----------------------
+
+In some cases, there may be members at ``np.core.*.*`` which we intended to be public-facing. Users of these will receive ``DeprecationWarning``s until we lift these members to ``np.core.*``.
```
Comment (rgommers) [inline doc/neps/nep-0023-restructuring-np-core.rst]: I don't think those exist. The whole `np.core` was never intended to be public (and arguably it isn't, missing underscores or not), so submodules of `np.core` certainly weren't intended to be public.

**Comment by rgommers** [comment]
I proposed something similar for scipy in 2011: https://mail.python.org/pipermail/scipy-dev/2011-June/016337.html. Didn't happen in the end.

File: doc/neps/nep-0023-restructuring-np-core.rst [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	 'shape_base',
+	 'umath']
+
+The publicness of this API surface is contagious - in principle, ``np.core.numerictypes.array`` (an alias of ``np.core.multiarray.array``) is part of the API too.
+
+This is problematic - it means that if ``numeric.py`` contains ``from numpy.core.multiarray import array``, then we can never remove that line, without risking breaking someone using ``np.core.numeric.array``.
+
+The proposal here is to:
+
+* Make all of the existing modules within `np.core` private, adding a leading underscore to their names.
```
Comment (rgommers) [inline doc/neps/nep-0023-restructuring-np-core.rst]: Maybe a lot less work: move `np.core` to `np._core` (and and import warnings for `np.core`)

File: doc/neps/nep-0023-restructuring-np-core.rst (hunk 1 above)
Comment (eric-wieser) [reply doc/neps/nep-0023-restructuring-np-core.rst]: Updated the text with this example

File: doc/neps/nep-0023-restructuring-np-core.rst (hunk 2 above)
Comment (charris) [reply doc/neps/nep-0023-restructuring-np-core.rst]: @rgommers A similar thought occurred to me. This is beginning to sound like the evolution of the `numpy.testing` package where we ended up with a couple of stub modules and `_private`. A clear statement of intent and justification would be good to have.

Comment (charris) [reply doc/neps/nep-0023-restructuring-np-core.rst]: I see that there is a statement of intent. Maybe we can expand this to a discussion of what we would like NumPy to look like in the long term.

**Comment by ahaldane** [comment]
I think I'm +1 on this. 

A clearly demarcated public/private api is a good thing. While maintaining back-compat is very important, so is dev freedom to change things, and sometimes it seems like the pendulum has swung to far to the former. It's not good to unnecessarly trap numpy into back-compat corners, as seems to have happened with these core modules.

The old modules can all be soft-deprecated, so no one's code will break except that they get deprecation warnings.

**Comment by mattip** [comment]
This is a NEP proposal, I think we should either move it forward as a draft NEP for wider disucssion or close the PR.

**Comment by rgommers** [comment]
This is a good idea, so please don't close it.

Move forward: @eric-wieser asked for a volunteer. I think this NEP is probably higher prio than most PRs that are open.

**Comment by mattip** [comment]
This NEP is still not merged to draft state, and still looking for someone to move it forward

File: doc/neps/nep-0023-restructuring-np-core.rst [hunk 3]
```diff
@@ -0,0 +1,165 @@
+==============================================
+NEP 23 — Making details of ``np.core`` private
+==============================================
```
Comment (charris) [inline doc/neps/nep-0023-restructuring-np-core.rst:3]: Number needs updating to 46. Also needs a new file name.

**Comment by charris** [comment]
I'm inclined to put this in for discussion, but needs updating.

**Comment by charris** [comment]
close/reopen.

File: doc/neps/nep-0023-restructuring-np-core.rst [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	 'machar',
+	 'multiarray',
+	 'numeric',
+	 'numerictypes',
+	 'rec',
+	 'records',
+	 'shape_base',
+	 'umath']
+
+The publicness of this API surface is contagious - in principle,
```
Comment (stefanv) [inline doc/neps/nep-0023-restructuring-np-core.rst:56]:
```suggestion
`np.core` and its submodules greatly increase the public API surface of NumPy—in principle, for example,
```

File: doc/neps/nep-0023-restructuring-np-core.rst [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+``np.core.numerictypes.array`` (an alias of ``np.core.multiarray.array``) is
+part of the API too.
+
+This is problematic - it means that if ``numeric.py`` contains ``from
+numpy.core.multiarray import array``, then we can never remove that line
+without risking breaking someone using ``np.core.numeric.array``.
```
Comment (stefanv) [inline doc/neps/nep-0023-restructuring-np-core.rst:62]:
```suggestion
without risking breaking code importing ``np.core.numeric.array``.
```

File: doc/neps/nep-0023-restructuring-np-core.rst [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  underscore to their names.
+* Introduce a set of new modules with the same names as the old ones, which:
+
+  * hard-code a list of the public API surface of their module at the time of
+    its creation
+  * emit a ``DeprecationWarning`` upon import, warning the user that they are
+    relying on implementation details.
+
+
+Implementation
```
Comment (stefanv) [inline doc/neps/nep-0023-restructuring-np-core.rst:76]: Maybe the implementation code should be left to a PR, that can be linked in.  My take was that the Implementation sections are for discussing implementation decisions, but not code.

File: doc/neps/nep-0023-restructuring-np-core.rst [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+type aliases.
+
+Backward compatibility
+----------------------
+
+In some cases, there may be members at ``np.core.*.*`` which we intended to be
+somewhat public-facing.  An example of such a function is
+`numpy.core.numeric.normalize_axis_index`, which downstream libraries are
+starting to use to validate axis arguments.  Users of these functions will
+receive ``DeprecationWarning``s until we lift these members to ``np.core.*``.
```
Comment (stefanv) [inline doc/neps/nep-0023-restructuring-np-core.rst:141]: Not sure what "until we lift these members to np.core" means here.  They should probably be lifted to `numpy.lib`?

File: doc/neps/nep-0023-restructuring-np-core.rst (hunk 3 above)
Comment (rgommers) [reply doc/neps/nep-0023-restructuring-np-core.rst:3]: 47 now

**Comment by rgommers** [comment]
I forgot about this NEP. Nice timing to comment on it @stefanv. I'll at least link to it from my upcoming NEP about the array API standard. I'd also be interested to push this forward, just a matter of finding the time .....

**Comment by rgommers** [comment]
Let me also link https://github.com/numpy/numpy/commits/master/numpy/tests/test_public_api.py, which has all the relevant public/private/in-between parts listed.

**Comment by seberg** [comment]
We discussed this today, and were landing a bit on "not really a big urgency".  This seems obviously a good idea, I personally do not even feel we need a NEP, unless there is a surprising amount of downstream breakage (but it cannot hurt).

The main concerns are:
* Where to put the symbols only available through core, this needs a concrete proposal to decide on.
* There may be quite a few PRs that have merge conflict.

Neither seem like big road blocks, and DeprecationWarnings can linger for a while (or even only be added after big downstream packages are fixed).

IMO, the main point here is to do the work, and if we want finish up the NEP in no particular order.

**Comment by rgommers** [comment]
+1 for this 

> There may be quite a few PRs that have merge conflict.

This was the case for the way we just did things in SciPy (adding back a `core.py` with a `__getattr__` and raising suitable deprecation warnings), but scikit-learn dynamically generates `core.py` and then there's no merge conflicts at all - Git will understand the `git mv` action.

> Where to put the symbols only available through core, this needs a concrete proposal to decide on.

Are there actually any that should be public? IIRC `numpy.core` submodules are not in the html docs, so there may not be any objects of interest? There's only very few examples given in the draft here, and they're pretty niche.

**Comment by seberg** [comment]
> There's only very few examples given in the draft here, and they're pretty niche.

Yes, I agree they are niche, but I suspect they exist.  But yes, even OK to deprecated and see who shouts, probably.

**Comment by WarrenWeckesser** [comment]
> Are there actually any that should be public?

At least one, already mentioned in the NEP:  `normalize_axis_index` from `numpy.core.multiarray` (also in `numpy.core.numeric`).  We use this in SciPy .  I asked about this on the [mailing list](https://www.mail-archive.com/numpy-discussion@python.org/msg03921.html) (note that I initially referred to the wrong namespace), and in that thread, @eric-wieser  said

>When I added this function, it was always my intent for it to be consumed
by downstream packages, but as Sebastian remarks, it wasn't really
desirable to put it in the top-level namespace.
>
> I think I would be reasonably happy to make the guarantee that it would not
be removed (or more likely, moved) without a lengthy deprecation cycle.

The comments in the thread suggest eventually moving the function to a "more public" namespace, e.g. `numpy.lib`, but not necessarily the top-level namespace.

FWIW, I also noticed the cupy issue https://github.com/cupy/cupy/issues/4522, and I see that dask uses `normalize_axis_index` and `normalize_axis_tuple` in the file [`numpy_compat.py`](https://github.com/dask/dask/blob/main/dask/array/numpy_compat.py), both imported from `numpy.core.numeric`.

**Pull request closed without merging**